Snackbar: Remove MarkupString capability from Add(string message…) and add Add(MarkupString message…) (#8146)

Conman-123 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->
`ISnackbar.Add()` accepts a string argument for the message content, but when rendering it casts this to a MarkupString to allow rendering of HTML. The only way to know this is to look in the documentation, so this opens it up to a XSS vulnerability.

This PR obsoletes this method and replaces it with an equivalent method that explicitly accepts a `MarkupString` as an argument instead of a `string`. Therefore all instances of `Snackbar.Add("Text")` had to be updated to be like `Snackbar.Add((MarkupString)"Text")`.

Resolves #8146.

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Updated all existing snackbar tests to use the new method, with the only change being to cast the input `string`s to `MarkupString`s.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [X] I've added relevant tests.

Conman-123 committed:
MudSnackbar: Obsolete Add(string message...) and replace with Add(MarkupString message...)

Conman-123 committed:
Merge branch 'MudBlazor:dev' into fix/snackbar-xss

Conman-123 commented:
My first PR, please let me know if I did anything wrong.

Should I add `[ExcludeFromCodeCoverage]` on the obsolete `Add(string)` method? I wasn't sure because some obsolete methods do and some don't.

Also wasn't sure if I should change the old `Add(string)` method to call the new `Add(MarkupString)` method, or just leave it to avoid any risk of breaking changes (since it is no longer unit tested).

henon commented:
@ScarletKuro, what do you think? Somehow I don't like that our nice simple examples now would have to cast to (MarkupString) everywhere. I would love to just undo the MarkupString feature of the `Add(string)` method in a breaking change but we would only be able to do that in v7. 

What if we obsolete `Add(string)` with a message that in v7 it will only render plain text not MarkupString? Then we could keep the examples all without a cast using the obsoleted method and use the MarkupString overload for the html example. Whereever the obsoleted message is used the warning can be ignored with pragmas.

I am not sure, so I am asking for a second opinion.

henon commented:
I guess what I propose would't counter the XSS vulerability :(

ScarletKuro commented:
> What if we obsolete Add(string) with a message that in v7 it will only render plain text not MarkupString?

Do you mean not to cast string to `MarkupString` and just show plain text so if there will be plain text even if html string is passed? That can solve the issue I guess.

henon commented:
Question is, do we break in v6 or v7? Is this vulnerability worth breaking in v6?

Another idea would be to just add an `AllowMarkupString` property to `MudGlobal` which people scared of XSS vulnerabilities can set to `false`. Then everywhere we support MarkupString we can disable that support by evaluating that global variable.

ScarletKuro commented:
Both solutions are fine with me. I'm more open to breaking change in the main-stream branch, since v7 future is too unknown.

henon commented:
> Both solutions are fine with me. I'm more open to breaking change in the main-stream branch, since v7 future is too unknown.

That makes sense, but it would be a shit way to break because it would be silent. No compilation error, no warning just suddenly html tags in your snackbar messages. 

@Conman-123 I mulled it over and I am sorry to say that even though I initially agreed I don't like deprecating the `Add(string )` method as it would force everyone to cast to `MarkupString` and I think 90% of users don't even use them. I'd rather add a global switch that allows everyone concerned with XSS vulnerabilities to just disable this feature. We can document the vulnerability and the switch in the Snackbar page in a warning box. 

Later on we can disable other uses of MarkupString in other components. Of if you want to do it in this PR you can disable all with this new switch.

Conman-123 commented:
No problem, I agree this wasn't a great solution but I couldn't think of anything better without a breaking change.
I'll give it a go when I get a chance.

danielchalmers commented:
@Conman-123 Work on v7 is underway in the dev branch so breaking changes aren't as much of a dealbreaker right now.

ScarletKuro commented:
Do we plan to do anything with that for v7?

Conman-123 commented:
@ScarletKuro for v7 we should add the breaking change to remove the silent cast of `string` to `Markupstring`, and have a separate overload that explicitly accepts a `Markupstring`. The fix with the config option for v6 can then be removed for v7.

Conman-123 commented:
@danielchalmers wait so breaking changes can be merged into dev now, and I will need to create a separate pull request for v6 with the non-breaking version of the fix?

danielchalmers commented:
> wait so breaking changes can be merged into dev now, and I will need to create a separate pull request for v6 with the non-breaking version of the fix?

Nope, everything goes straight to v7 and we don't worry about v6 right now! Much easier 😄 

(if someone wants a feature or bugfix backported to v6, that's their responsibility)

henon commented:
@Conman-123 please continue this PR if you will. This is what we do:

We keep `Add(string)` but it does not use `MarkupString` and we add the overload `Add(MarkupString)`. Marking this as breaking change for v7

Commit:
Revert "MudSnackbar: Obsolete Add(string message...) and replace with Add(MarkupString message...)"

This reverts commit bebc3ee20ea75f3d1f338959cd954c73545cd1c1.

Commit:
Merge branch 'dev' into fix/snackbar-xss

Commit:
- [Breaking Change] Change Snackbar.Add(string...) so it no longer converts to MarkupString and renders HTML
- Add Snackbar.Add(MarkupString...)
- Add test to ensure MarkupString works
- Add test to ensure string is not rendered as HTML

Commit:
Add more tests

Conman-123 commented:
@henon I've updated this pull request according to your suggestions

henon commented:
Added to v7.0.0 Migration Guide #8447. @ScarletKuro please merge, if all is in order.

ScarletKuro commented:
Thanks
